### ISSUE TITLE
UnregisterClientHandler dont work 

### ISSUE THREAD

**Issue by yxlong** [b0ykoe/ducksoup#14]
Hi b0ykoe,

i use ducksoup develop version.
the fun `UnregisterClientHandler `  in PacketHandler.cs dont work when i invoke it

When I debug, I found that the data returned by this code snippet is empty: `var keysToRemove = _clientHandlers[msgId].Where(m => m.Value.Equals(handler)).Select(c => c.Key).ToList();`. I think the Equals method might not be working.

**Comment by b0ykoe** [comment]
Hey!

Can you give me a example on how you're using it to register and unregister handlers?

**Comment by yxlong** [comment]
OK。i register handler in Plugin,like the example you gived.
When Plugin load , it's work.
But the handler register  in Plugin always work when Plugin be unloaded
so i debug the UnregisterClientHandler,found the code stop at   the line (`var keysToRemove = _clientHandlers[msgId].Where(m => m.Value.Equals(handler)).Select(c => c.Key).ToList();`)

i think problem is the args handle type is diffrent from handle in _clientHandlers. 

i fixed it , but maybe my code is not good.
here is my code replace UnregisterClientHandler  :
1、change `private readonly Func<T, ISession, Task<Packet>> _handlerAction`  be `public readonly Func<T, ISession, Task<Packet>> _handlerAction;`
2、change code 

```
public void UnregisterClientHandler<T>(Func<T, ISession, Task<Packet>> handler) where T : Packet, new()
{
    var msgId = new T().MsgId;
    if (_clientHandlers.GetValueOrDefault(msgId, null) == null) return;
    var temp = _clientHandlers[msgId];   
    var keysToRemove = temp.Where(m => (m.Value as _PacketHandler<T>)._handlerAction.Equals(handler)).Select(c => c.Key).ToList();//here
    
    foreach ( var keyValuePair in keysToRemove)
    {
        Console.WriteLine(keyValuePair);
    }
    keysToRemove.ForEach(key => _clientHandlers[msgId].Remove(key, out var tempObject));
}
```

**Comment by b0ykoe** [comment]
Hi, thanks for pointing out this issue. 

I've implemented some changes for it to work properly. This was fixed in ce3de591ad2f6fd153f201cd44f880ccbf39c638

Short example:

```csharp
        PacketHandler.RegisterClientHandler<CLIENT_CHAT_REQUEST>(TestHandler);

    public async Task<Packet> TestHandler(CLIENT_CHAT_REQUEST data, ISession session)
    {
        if (data.Message.ToLower().Contains("abc"))
        {
            session.SendNotice("Hi!");
            data.ResultType = PacketResultType.Block;
            return data;
        }

        if (data.Message.ToLower().Contains("123"))
        {
            session.SendNotice("Trying to unregister");
            Task.Run(() =>
            {
                PacketHandler.UnregisterClientHandler<CLIENT_CHAT_REQUEST>(TestHandler);
            });
            data.ResultType = PacketResultType.Block;
            return data;
        }
        
        return data;
    }
```
